File: PCM_Mobile/lib/models/booking.dart
import 'enums.dart';

class Booking {
  final int id;
  final int courtId;
  final int memberId;
  final DateTime startTime;
  final DateTime endTime;
  final double totalPrice;
  final BookingStatus status;
  final String? recurrenceRule;
  final DateTime createdDate;
  final String? courtName;
  final String? memberName;

  Booking({
    required this.id,
    required this.courtId,
    required this.memberId,
    required this.startTime,
    required this.endTime,
    required this.totalPrice,
    required this.status,
    this.recurrenceRule,
    required this.createdDate,
    this.courtName,
    this.memberName,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    // SAFE ENUM PARSING: Handle out-of-range enum values
    BookingStatus parseStatus(dynamic value) {
      if (value is int && value >= 0 && value < BookingStatus.values.length) {
        return BookingStatus.values[value];
      }
      return BookingStatus.pendingPayment; // Default fallback
    }

    return Booking(
      id: json['id'],
      courtId: json['courtId'],
      memberId: json['memberId'],
      startTime: DateTime.parse(json['startTime']),
      endTime: DateTime.parse(json['endTime']),
      totalPrice: (json['totalPrice'] as num).toDouble(),
      status: parseStatus(json['status']),
      recurrenceRule: json['recurrenceRule'],
      createdDate: json['createdDate'] != null 
          ? DateTime.parse(json['createdDate']) 
          : DateTime.now(),
      courtName: json['courtName'],
      memberName: json['memberName'],
    );
  }
}

File: PCM_Backend/PCM.API/Models/Booking.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace PCM.API.Models;

[Table("519_Bookings")]
public class Booking
{
    [Key]
    public int Id { get; set; }
    
    public int CourtId { get; set; }
    
    public int MemberId { get; set; }
    
    public DateTime StartTime { get; set; }
    
    public DateTime EndTime { get; set; }
    
    [Column(TypeName = "decimal(18,2)")]
    public decimal TotalPrice { get; set; }
    
    public int? TransactionId { get; set; }
    
    public bool IsRecurring { get; set; } = false;
    
    [MaxLength(200)]
    public string? RecurrenceRule { get; set; }
    
    public int? ParentBookingId { get; set; }
    
    public BookingStatus Status { get; set; } = BookingStatus.PendingPayment;
    
    public DateTime CreatedDate { get; set; } = DateTime.Now;
    
    public DateTime? HoldExpiresAt { get; set; }
    
    // Navigation
    [ForeignKey(nameof(CourtId))]
    public virtual Court Court { get; set; } = null!;
    
    [ForeignKey(nameof(MemberId))]
    public virtual Member Member { get; set; } = null!;
}
